File: Leetcode_Python/median_of_two_sorted_arrays.py
'''Version 1: O(log(m + n))'''
class Solution:
    # @return a float
    def findMedianSortedArrays(self, A, B):
        length = len(A) + len(B)
        if length % 2 == 1:
            return self.findKth(A, B, length / 2 + 1)
        else:
            return (self.findKth(A, B, length / 2) + self.findKth(A, B, length / 2 + 1)) / 2.0
        
    def findKth(self, A, B, k):
        if len(A) > len(B):
            return self.findKth(B, A, k)
        else:
            if len(A) == 0:
                return B[k - 1]
            if k == 1:
                return min(A[0], B[0])
            else:
                pa = min(k / 2, len(A))
                pb = k - pa
                if A[pa - 1] <= B[pb - 1]:
                    return self.findKth(A[pa:], B, k - pa)
                else:
                    return self.findKth(A, B[pb:], k - pb)

'''Version 2: O(m + n)'''
class Solution:
    # @return a float
    def findMedianSortedArrays(self, A, B):
        if not A:
            return B[len(B)/2] if len(B) % 2 == 1 else (B[len(B) / 2 - 1] + B[len(B) / 2]) / 2.0
        if not B:
            return A[len(A)/2] if len(A) % 2 == 1 else (A[len(A) / 2 - 1] + A[len(A) / 2]) / 2.0
        i = j = 0
        C = []
        while i < len(A) and j < len(B):
            if A[i] < B[j]:
                C.append(A[i])
                i += 1
            else:
                C.append(B[j])
                j += 1
        while i < len(A):
            C.append(A[i])
            i += 1
        while j < len(B):
            C.append(B[j])
            j += 1
        k = len(C)
        return C[k/2] if k % 2 == 1 else (C[k / 2 - 1] + C[k / 2]) / 2.0
        

                

File: Leetcode_Java/MedianOfTwoSortedArrays.java
public class MedianOfTwoSortedArrays {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        int length = nums1.length + nums2.length;
        if (length % 2 == 1) {
            return findKthElement(nums1, 0, nums1.length - 1,
                nums2, 0, nums2.length - 1, length / 2 + 1);
        } else {
            return (findKthElement(nums1, 0, nums1.length - 1, nums2, 0,
                nums2.length - 1, length / 2) + findKthElement(nums1,
                0, nums1.length - 1, nums2, 0, nums2.length - 1, length / 2 + 1)) / 2.0;
        }
    }

    private int findKthElement(int[] nums1, int oneStart, int oneEnd,
        int[] nums2, int twoStart, int twoEnd, int k) {
        int len1 = oneEnd - oneStart + 1;
        int len2 = twoEnd - twoStart + 1;
        if (len1 < len2) {
            return findKthElement(nums2, twoStart, twoEnd, nums1, oneStart, oneEnd, k);
        }
        if (len2 == 0) { // this case should be dealt before case 2
                         // to avoid ArrayIndexOutOfBoundsException
            return nums1[oneStart + k - 1];
        }
        if (k == 1) { // case 2
            return Math.min(nums1[oneStart], nums2[twoStart]);
        }
        int mid2 = Math.min(k / 2, len2);
        int mid1 = k - mid2;
        if (nums1[oneStart + mid1 - 1] > nums2[twoStart + mid2 - 1]) {
            return findKthElement(nums1, oneStart, oneEnd, nums2,
                twoStart + mid2, twoEnd, k - mid2);
        } else {
            return findKthElement(nums1, oneStart + mid1, oneEnd, nums2,
                twoStart, twoEnd, k - mid1);
        }
    }

    public static void main(String[] args) {
        MedianOfTwoSortedArrays test = new MedianOfTwoSortedArrays();
        int[] nums1 = {1};
        int[] nums2 = {3, 4};
        System.out.println(test.findMedianSortedArrays(nums1, nums2));
    }
}